cache/repo_cache_bug.go

```diff
@@ -108,36 +111,72 @@ func (c *RepoCache) writeBugCache() error {
 func (c *RepoCache) ResolveBugExcerpt(id entity.Id) (*BugExcerpt, error) {
 	c.muBug.RLock()
 	defer c.muBug.RUnlock()
-
-	e, ok := c.bugExcerpts[id]
+	excerpt, ok := c.bugExcerpts[id]
 	if !ok {
-		return nil, bug.ErrBugNotExist
+		return nil, errBugNotInCache
 	}
-
-	return e, nil
+	c.presentBugs.Get(id)
+	return excerpt, nil
 }
 
 // ResolveBug retrieve a bug matching the exact given id
 func (c *RepoCache) ResolveBug(id entity.Id) (*BugCache, error) {
+	err := c.ensureBugLoaded(id)
+	if err != nil {
+		return nil, err
+	}
+
 	c.muBug.RLock()
-	cached, ok := c.bugs[id]
-	c.muBug.RUnlock()
-	if ok {
-		return cached, nil
+	defer c.muBug.RUnlock()
+	return c.bugs[id], nil
+}
+
+// ensureBugLoaded ensures a bug (if it exists) is loaded into the cache
+// it will automatically evict a bug if needed
+func (c *RepoCache) ensureBugLoaded(id entity.Id) error {
+	if c.presentBugs.Get(id) {
+		return nil
 	}
 
 	b, err := bug.ReadLocalBug(c.repo, id)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	cached = NewBugCache(c, b)
+	bugCache := NewBugCache(c, b)
+
+	err = c.evictIfNeeded()
+	if err != nil {
+		return err
+	}
 
 	c.muBug.Lock()
-	c.bugs[id] = cached
-	c.muBug.Unlock()
+	defer c.muBug.Unlock()
+	c.presentBugs.Add(id)
+	c.bugs[id] = bugCache
+	return nil
+}
 
-	return cached, nil
+// evictIfNeeded will evict a bug from the cache if needed
+// it also removes references of the bug from the bugs and bugExcerpts
+func (c *RepoCache) evictIfNeeded() error {
+	c.muBug.Lock()
+	defer c.muBug.Unlock()
+	if c.presentBugs.Len() < c.presentBugs.maxSize {
+		return nil
+	}
+
+	for _, id := range c.presentBugs.GetAll() {
+		if b := c.bugs[id]; !b.NeedCommit() {
+			b.mu.Lock()
+			c.presentBugs.Remove(id)
+			delete(c.bugExcerpts, id)
```

Comment:
evicting from the cache should not change `c.bugExcerpts` as those as the full collection of bug excerpt to make search on it.